calour/io.py

```diff
@@ -80,13 +84,50 @@ def _read_biom(fp, transpose=True):
     return sid, fid, data, feature_md
 
 
-def _read_qiime2(fp, transpose=True):
-    '''Read in a qiime2 .qza biom table file.
+def _file_from_zip(tempdir, data_file, internal_data):
+    '''extract the data file from a regular/qza filename into the tempdir, and return the filename
+
+    Parameters
+    ----------
+    tmpdir: str
+        name of the directory to extract the zip into
+    data_file: str
+        original name of the file (could be '.qza' or not)
+    internale_data: str
+        the internal qiime2 qza file name (i.e. 'data/feature-table.biom' for biom table etc.)
+
+    Returns
+    -------
+    str: name of data file inside the tempdir
+    '''
+    if not zipfile.is_zipfile(data_file):
+        raise ValueError('the qiime2 file %s is not a valid zip file. Is it a qiime2 artifact (.qza) file?' % data_file)
+    fl = zipfile.ZipFile(data_file)
+    internal_name = None
+    for fname in fl.namelist():
+        if fname.endswith(internal_data):
```

Comment:
this should be `==` instead of `endswith`?

Reply:
btw, you should use the following?

````
with ZipFile('spam.zip') as myzip:
    with myzip.open('eggs.txt') as myfile:
        print(myfile.read())
````

Reply:
we cannot use the .open() because io.read() expects a file name and not a file.
so we need to unzip to a temp dir and then use io.read to load it

Reply:
endswith since qiime2 includes a random/hashed subfolder name in its zipfile
cannot use open() since biom.parse_table() uses the file.tell() method, which is supported for zipfile only in python 3.7 and not in 3.6 so parse_table fails.